File: app/src/main/java/com/anthonysottile/kenken/ui/KenKenSquare.java
package com.anthonysottile.kenken.ui;

import android.graphics.Canvas;
import android.graphics.Paint;

import com.anthonysottile.kenken.UserSquare;

import java.util.ArrayList;
import java.util.EventListener;
import java.util.EventObject;
import java.util.List;

public class KenKenSquare {

    public enum SquareTouchState {
        None,
        Touching,
        Selected
    }

    private UserSquare square;

    public UserSquare getUserSquare() {
        return this.square;
    }

    private boolean markedIncorrect = false;

    public void setMarkedIncorrect(boolean markedIncorrect) {
        this.markedIncorrect = markedIncorrect;
        this.triggerRequestRedrawEvent();
    }

    private SquareTouchState touchState = SquareTouchState.None;

    public SquareTouchState getTouchState() {
        return this.touchState;
    }

    public void setTouchState(SquareTouchState touchState) {
        this.touchState = touchState;
        this.triggerRequestRedrawEvent();
    }

    private String cageText = "";

    public void setCageText(String cageText) {
        this.cageText = cageText;
        this.triggerRequestRedrawEvent();
    }

    public void drawSquare(Canvas canvas, SquareDrawingDimensions dimensions) {
        int x = this.square.getX();
        int y = this.square.getY();

        // Fill background
        Paint backgroundColor;
        switch (this.touchState) {
            case Touching:
                backgroundColor = UIConstants.GetHoveringColor();
                break;
            case Selected:
                backgroundColor = UIConstants.GetSelectedColor();
                break;
            case None:
            default:
                backgroundColor = UIConstants.GetBackgroundColor();
                break;
        }

        if (this.markedIncorrect) {
            backgroundColor = UIConstants.GetMarkedIncorrectColor();
        }

        dimensions.PaintBackgroundColor(
                canvas,
                backgroundColor,
                x,
                y
        );

        // Draw cage text
        if (this.cageText.length() > 0) {
            dimensions.PaintCageText(
                    canvas,
                    this.cageText,
                    x,
                    y
            );
        }

        // Draw value text or candidates text
        int squareValue = this.square.getValue();
        if (squareValue > 0) {
            // Display value
            dimensions.PaintValueText(
                    canvas,
                    Integer.toString(squareValue),
                    x,
                    y
            );
        } else {
            // Display candidates
            String candidatesText = this.square.getCandidatesString();

            // Only paint if there is a string there
            if (candidatesText.length() > 0) {
                dimensions.PaintCandidatesText(
                        canvas,
                        candidatesText,
                        x,
                        y
                );
            }
        }
    }

    public KenKenSquare(UserSquare square) {
        this.square = square;

        this.square.addChangedEventHandler(new UserSquare.UserSquareChangedListener() {
            public void onUserSquareChanged(EventObject event) {
                KenKenSquare.this.triggerRequestRedrawEvent();
            }
        });
    }

    // #region Request Redraw Event

    public interface RequestRedrawListener extends EventListener {
        void onRequestRedraw(EventObject event);
    }

    private final List<RequestRedrawListener> requestRedrawListeners =
            new ArrayList<RequestRedrawListener>();

    public void AddRequestRedrawListener(RequestRedrawListener listener) {
        this.requestRedrawListeners.add(listener);
    }

    public void RemoveRequestRedrawListener(RequestRedrawListener listener) {
        this.requestRedrawListeners.remove(listener);
    }

    public void ClearRequestRedrawListeners() {
        this.requestRedrawListeners.clear();
    }

    private void triggerRequestRedrawEvent() {
        EventObject event = new EventObject(this);

        for (RequestRedrawListener listener : this.requestRedrawListeners) {
            listener.onRequestRedraw(event);
        }
    }

    // #endregion

}

File: app/src/main/java/com/anthonysottile/kenken/ui/KenKenSquare.kt
package com.anthonysottile.kenken.ui

import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import com.anthonysottile.kenken.UserSquare
import java.util.*

class KenKenSquare(val userSquare: UserSquare) {

    enum class SquareTouchState {
        None,
        Touching,
        Selected
    }

    var markedIncorrect = false
        set(markedIncorrect) {
            field = markedIncorrect
            this.triggerRequestRedrawEvent()
        }

    var touchState = SquareTouchState.None
        set(touchState) {
            field = touchState
            this.triggerRequestRedrawEvent()
        }

    var cageText = ""
        set(cageText) {
            field = cageText
            this.triggerRequestRedrawEvent()
        }

    private val requestRedrawListeners = ArrayList<() -> Unit>()

    fun drawSquare(canvas: Canvas, dimensions: SquareDrawingDimensions, x: Int, y: Int) {
        // Fill background
        var backgroundColor: Paint = when (this.touchState) {
            KenKenSquare.SquareTouchState.Touching -> hoveringColor
            KenKenSquare.SquareTouchState.Selected -> selectedColor
            KenKenSquare.SquareTouchState.None -> defaultColor
        }
        if (this.markedIncorrect) {
            backgroundColor = incorrectColor
        }

        dimensions.paintBackgroundColor(
                canvas,
                backgroundColor,
                x,
                y
        )

        // Draw cage text
        if (this.cageText.isNotEmpty()) {
            dimensions.paintCageText(
                    canvas,
                    this.cageText,
                    x,
                    y
            )
        }

        // Draw value text or candidates text
        val squareValue = this.userSquare.value
        if (squareValue > 0) {
            // Display value
            dimensions.paintValueText(
                    canvas,
                    Integer.toString(squareValue),
                    x,
                    y
            )
        } else {
            // Display candidates
            val candidatesText = this.userSquare.getCandidatesString()

            // Only paint if there is a string there
            if (candidatesText.isNotEmpty()) {
                dimensions.paintCandidatesText(
                        canvas,
                        candidatesText,
                        x,
                        y
                )
            }
        }
    }

    init {
        this.userSquare.addChangedEventHandler { _ -> this@KenKenSquare.triggerRequestRedrawEvent() }
    }

    fun addRequestRedrawListener(listener: () -> Unit) {
        this.requestRedrawListeners.add(listener)
    }

    private fun triggerRequestRedrawEvent() {
        for (listener in this.requestRedrawListeners) {
            listener()
        }
    }

    companion object {
        private val hoveringColor = Paint()
        private val selectedColor = Paint()
        private val incorrectColor = Paint()
        private val defaultColor = Paint()

        init {
            hoveringColor.color = Color.rgb(0x1e, 0x77, 0xd3)
            selectedColor.color = Color.rgb(0x00, 0xc0, 0xc0)
            incorrectColor.color = Color.rgb(0xff, 0x66, 0x66)
            defaultColor.color = Color.rgb(0xff, 0xff, 0xff)
        }
    }
}
